### DISCUSSION TITLE
Cannot build skia-bindings from crates.io when building Skia from source and binary cache disabled

### DISCUSSION THREAD

**Discussion by tronical** [rust-skia/rust-skia#665] [General]
The following steps reproduce this issue:

```sh
cargo new skia-test
cd skia-test
cargo add skia-safe --no-default-features --features gl
cargo build
```

Error:

```sh
error: failed to run custom build command for `skia-bindings v0.52.0`

Caused by:
  process didn't exit successfully: `/Users/simon/skia-test/target/debug/build/skia-bindings-62dd4d0a97f016e6/build-script-build` (exit status: 101)
  --- stdout
  cargo:rerun-if-env-changed=SKIA_DEBUG
  cargo:rerun-if-env-changed=SKIA_SOURCE_DIR
  STARTING A FULL BUILD
  HOST: x86_64-apple-darwin
  cargo:rerun-if-env-changed=OPT_LEVEL
  cargo:rerun-if-env-changed=CC
  cargo:rerun-if-env-changed=CXX
  cargo:rerun-if-env-changed=SKIA_USE_SYSTEM_LIBRARIES
  cargo:rerun-if-env-changed=SDKROOT
  cargo:rerun-if-env-changed=MACOSX_DEPLOYMENT_TARGET
  cargo:rerun-if-env-changed=SKIA_NINJA_COMMAND
  cargo:rerun-if-env-changed=SKIA_GN_COMMAND
  Probing 'python'
  Probing 'python3'
  Python 3 found: "python3"
  Synchronizing Skia dependencies

  --- stderr
  /usr/local/bin/python3: can't open file '/Users/simon/.cargo/registry/src/github.com-1ecc6299db9ec823/skia-bindings-0.52.0/skia/tools/git-sync-deps': [Errno 2] No such file or directory
  thread 'main' panicked at '`skia/tools/git-sync-deps` failed', /Users/simon/.cargo/registry/src/github.com-1ecc6299db9ec823/skia-bindings-0.52.0/build_support/skia/config.rs:370:9
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

and indeed a look into `.cargo/registry/src/github.com-1ecc6299db9ec823/skia-bindings-0.52.0/` yields no `skia` sub-directory.

Enabling the `binary-cache` features appears to correctly download the sources (when `FORCE_SKIA_BUILD=1` or otherwise no cached binaries).

This also works when using `skia-safe` from git (and this `skia-bindings` implicitly from git, too). I guess that's because it's a sub-module then.

**Comment by pragmatrix** [comment]
Hi @tronical,

The crate version of skia-bindings is intended for building from binaries or building from downloaded sources as a fallback. When skia-bindings attempts to download either binaries or sources, the `binary-cache` feature must be enabled. If the `binary-cache` feature is disabled, no download attempt will be made, meaning that the sources must be made available somewhere else and the `SKIA_SOURCE_DIR` and optionally the `SKIA_NINJA_COMMAND` and `SKIA_GN_COMMAND` environment variables should be used to specify where they are.

What do you think should happen here? And what is your actual use-case for using the crate version?

**Comment by tronical** [comment]
Looks like I misunderstood the setup indeed.

The background is that for [Slint](https://github.com/slint-ui/slint) we want to offer to our users the option of using Skia for rendering.

Your bindings make that really easy for us - great job :)

Since we upload to crates.io we would like to depend on a released version of your bindings.

However it is quite likely that the binary cache is not available or suitable, for example when building against a Yocto SDK.

In that case I was - erroneously - hoping that the bindings would build from source out of the box.

Is a download of the sources a feature you would consider accepting? (I might give it a try)

Or would you rather suggest that we try to build against a system installed Skia?

**Comment by pragmatrix** [comment]
So far I don't understand why the default behavior of `binary-cache` does not work for your usecase. Again, if binaries are not available, the `binary-cache` feature falls back downloading sources and builds Skia from them. The situations for which that is not a good default is when offline builds or custom Skia patches are needed.

> Is a download of the sources a feature you would consider accepting? (I might give it a try)

Sure, but what exactly would be different compared to the default behavior of the `binary-cache` feature?

> Or would you rather suggest that we try to build against a system installed Skia?

No. Depending on the features you enable, that would be rather difficult. The bindings are tailored to the exact Skia milestone and we added [a number of patches](https://github.com/google/skia/compare/chrome/m104...rust-skia:m104-0.52.1) to Skia so that it builds all configurations we support.

**Reply by tronical** [reply]
You’re right, the binary-cache feature is the way to go. Thanks for the explanations (and conversion to GH discussion)

I have one build issue left with binary-cache, but that’s not related to the feature as-is and I need to solve it separately. (It’s basically when cross compiling in a yocto environment the additional build dependencies the feature pulls in for the download itself doesn’t compile)

**Comment by antonsmetanin** [comment]
I have the same issue and I'm not sure I understand how it's supposed to work.
I assumed that disabling the `binary-cache` feature would skip using prebuilt libraries and fall back to downloading the source files and building the libraries from them. But instead it skips both and fails with the aforementioned error (`skia/tools/git-sync-deps failed`).

My use case for this is that prebuilt binaries cause linker errors in `std::find` and `std::max_element` (which I don't really know how to fix):
```
= note: skia.lib(skia.SkSLVMCodeGenerator.obj) : error LNK2019: unresolved external symbol __std_max_element_8 referenced in function "private: struct SkSL::`anonymous namespace'::Value __cdecl SkSL::SkVMGenerator::writeIntrinsicCall(class SkSL::FunctionCall const &)" (?writeIntrinsicCall@SkVMGenerator@SkSL@@AEAA?AUValue@?A0x6BA93B68@2@AEBVFunctionCall@2@@Z)
          skia.lib(gpu.GrGLCaps.obj) : error LNK2019: unresolved external symbol __std_find_trivial_4 referenced in function "public: bool __cdecl GrGLCaps::programBinaryFormatIsValid(unsigned int)const " (?programBinaryFormatIsValid@GrGLCaps@@QEBA_NI@Z)
          skia.lib(skia.SkSLFindAndDeclareBuiltinFunctions.obj) : error LNK2019: unresolved external symbol __std_find_trivial_8 referenced in function "void __cdecl SkSL::Transform::FindAndDeclareBuiltinFunctions(struct SkSL::Program &)" (?FindAndDeclareBuiltinFunctions@Transform@SkSL@@YAXAEAUProgram@2@@Z)
          skia.lib(skia.SkSLFindAndDeclareBuiltinVariables.obj) : error LNK2001: unresolved external symbol __std_find_trivial_8
          skia.lib(skia.SkNWayCanvas.obj) : error LNK2001: unresolved external symbol __std_find_trivial_8
```

**Reply by pragmatrix** [reply]
> error LNK2019: unresolved external symbol

I think I saw the same errors a few days ago, which disappeared after Visual Studio Build tools and / or Visual Studio were updated to the latest version. May be that helps.

>I assumed that disabling the binary-cache feature would skip using prebuilt libraries and fall back to downloading the source files and building the libraries from them. 

With `binary-cache` disabled, no download will be attempted, not even the sources. The feature was introduced to remove a number of dependencies that are needed to download files and which had build problems on some platforms. See #525. When this feature gets disabled in a crate build, skia-bindings expects the sources provided through the `SKIA_SOURCE_DIR` environment variable. 

To resolve the justified confusion, it would probably a be a good idea to add another default feature that enables the download and build from sources.

**Reply by antonsmetanin** [reply]
Okay, I see. Thanks, I'll try to update VS Build tools and see if it solves the problem.

I agree, it would be nice to split the feature into two separate ones. Or at least produce a more descriptive error with the current behaviour.
